Add platforms to the new project wizard

stevemessick opened the pull request:
Second page of the Android Studio 4.2 wizard looks like this. @devoncarew @pq may have ideas for improvement.

<img width="978" alt="Screen Shot 2020-11-20 at 3 50 54 PM" src="https://user-images.githubusercontent.com/8518285/99860538-3b738c00-2b48-11eb-9c46-68af1ee67be5.png">

The IntelliJ view is similar but includes a Project Type selector.

Instead of running `flutter config` this reads the settings file directly. It is much faster. Also, instead of running `flutter channel` this runs the git command directly.

We need to add unit tests to `flutter_tools` that document which fields we depend on from the settings file. We should also have a test that warns `flutter_tools` authors that if the locations of the settings file changes that IDE authors need to be notified.

The check boxes still need to be added to the wizard used by AS 4.1. That is a separate project, which may resolve itself when 4.2 becomes stable.

Line 26 of CONTRIBUTING.md documents the additional dependency on git4idea.

stevemessick commented:
One of the `AsyncRateLimiter` tests has gotten flakey again. Ignore it for now. I wonder if it is due to the speed of the Mac it runs on. Tests run in half the time as compared to previous CI.

devoncarew reviewed (commented):
I have some in-line comments; but here are broader thoughts:

If we don't have a provisioned SDK (if 'flutter anything' has not yet been run and the sdk hasn't set itself up), we could do something like just assume every platform is available.

Or, we could have the previous page in the sdk wizard ensure that the sdk is installed and has been set up; we could show a progress indicator if this is long running. That would allow later wizard pages to assume that every sdk operation would be fast.

If sdk operations are fast, or the sdk has been installed and set up, we could then rely on either querying the flutter tool for supported and legal platforms, or, have the flutter tool write a json file in a known location that we can read. That way we won't have to reply on in-lining flutter tool logic (config file location and format, use of git branches, ...) into the IntelliJ plugin.

stevemessick commented on resources/META-INF/plugin.xml:
```diff
@@ -24,6 +24,7 @@
   <idea-version since-build="201.7223" until-build="202.8194.7"/>
 
   <depends>Dart</depends>
+  <depends>Git4Idea</depends>
```
Yes

stevemessick commented on src/io/flutter/sdk/FlutterSdk.java:
```diff
@@ -474,12 +497,85 @@ public String getDartSdkPath() {
+  @NonNls
+  public Set<String> queryConfiguredPlatforms(boolean useCachedValue) {
+    Set<String> platforms = new HashSet<>();
+    // Someone could do: flutter config --no-enable-ios --no-enable-android
```
It doesn't. But if someone went to that much trouble we should initialize the checkboxes to respect the setting, I think.

stevemessick commented on src/io/flutter/FlutterBundle.properties:
```diff
@@ -211,3 +211,16 @@ settings.enable.embedding.devtools=Enable embedding DevTools in the Flutter Insp
+npw_platform_macos=MacOS
+npw_platform_web=Web
+npw_platform_windows=Windows
+npw_platform_availability_help=Platform availability depends on the Flutter SDK channel, and on 'flutter config'.
```
I wasn't sure how verbose we should be. I considered adding another line with some explanation, and think I'll do that now. However, I misunderstood the purpose of the config file. I'm going to have to rewrite some of this. Instead of enabling check boxes based on platforms enabled in the config file it seems those should be the ones checked by default. All should be enabled.

stevemessick commented:
> If we don't have a provisioned SDK ...

I wasn't sure which way to go with this. If we have an unusual configuration (incomplete SDK, or on a non-standard branch) do we display all platforms, or do we treat it like stable and hide the check boxes? The current implementation does the latter. That might be unfriendly to flutter engineers who are using a private branch. Having your own branch is probably a signal that you're workin in the master channel.

> That way we won't have to reply on in-lining flutter tool logic (config file location and format, use of git branches, ...) into the IntelliJ plugin.

I discussed this with some people on the Flutter team. This was their suggestion. However, they also said to add tests to flutter_tools that ensure if any of the inlined values change that a failing test would cause that engineer to notify IDE extension authors of the change. I'm working on that now, and don't plan to merge this until it's at least up for review. The flutter commands take 1-4 seconds in my testing. Using them noticeably slows down the initial appearance of the NPW.

stevemessick commented:
PTAL. I added more text.

<img width="985" alt="Screen Shot 2020-12-11 at 3 32 28 PM" src="https://user-images.githubusercontent.com/8518285/101965055-3c885e00-3bc8-11eb-9243-3b5f3374b22d.png">

It turns out that the desktop platforms must be enabled by `flutter config` before they will be generated, but andoird, ios, and linux do not. So I changed the enablement to reflect that.

stevemessick commented:
The latest commit removes the code to read the config file and runs `flutter config` instead. It was not practical to write the test required to keep the config file in sync. The `flutter_tools` test framework does not provide an override for `Cache` so a new file system is required. I ran into https://github.com/google/file.dart/issues/140 trying to implement one. The work-around for that is to write a new file system, a bit much for a consistency-checking test.

Initial display of the NPW wizard is a bit slower now, but we're going to have to live with that for a while.

devoncarew reviewed (approved):
The new wizard text looks great - thanks!

> The latest commit removes the code to read the config file and runs flutter config instead. It was not practical to write the test required to keep the config file in sync. The flutter_tools test framework does not provide an override for Cache so a new file system is required. I ran into google/file.dart#140 trying to implement one. The work-around for that is to write a new file system, a bit much for a consistency-checking test.
> Initial display of the NPW wizard is a bit slower now, but we're going to have to live with that for a while.

Gotcha. I think we should evolve the wizard code - in future PRs - so that the initial wizard page both allows you to select a Flutter SDK and makes sure that that SDK is warmed up - that everything necessary for `flutter config` / `flutter --version` / `flutter create` to run fast has been downloaded by the SDK.

stevemessick commented:
Agreed. I'm planning to do that in the work that adds the `Detect SDK` button we discussed recently. Just to recap that:
- Add a link to the Flutter installation web page
- Add a button to detect SDK
- Ensure the SDK is ready to use before transitioning to the next page
  - that may not be possible to enforce due to limitations of the wizard framework
  - the "warm up" will probably be modal, needing progress indication, so will be somewhat self-enforcing
